mautrix_signal/portal.py

```diff
@@ -1677,11 +1677,16 @@ async def _update_participants(self, source: u.User, info: ChatInfo) -> None:
                         self.mxid, user.mxid, "You are not a member of this Signal group"
                     )
 
-    async def _update_power_levels(self, info: ChatInfo) -> None:
+    async def _update_power_levels(self, info: ChatInfo, sender: p.Puppet | None = None) -> None:
         if not self.mxid:
             return
 
         power_levels = await self.main_intent.get_power_levels(self.mxid)
+        if sender:
+            power_levels = await self._get_power_levels(
+                power_levels, info=info, is_initial=False, sender=sender
+            )
+            await sender.intent_for(self).set_power_levels(self.mxid, power_levels)
```

Comment:
I think you want an early return here or an `else`, right?

Reply:
If there is a sender, `set_power_levels` must be run twice, once setting all power levels the puppet is allowed to send and again to set those the puppet isn't allowed to send by the bridge bot. Demotions from ADMINISTRATOR to DEFAULT for example can not be performed by the sender, since the sender and the demotee have the same power level (this is allowed on signal). In this case `if sender` will be executed but produce no change, the change must then be performed by the bridge bot.
What I mentioned in the PR description is that this runs set_power_levels unnecessarily quite a lot (usually either the command by the puppet or the command by the bridge bot will produce no change), so we might want to detect whether anything has changed before sending the request. I'm not really sure about a good way to do this, returning a `bool: changed` with `_get_power_levels` seems to be quite complicated, maybe we can pass a deep copy to `_get_power_levels` and compare before and after?